### PR TITLE
RM-7417: Nullable Reference Types in Core.Reflection

### PR DESCRIPTION
**Pull request by dkudernatsch**

### CODE REVIEW

File: Remotion/Core/Reflection.CodeGeneration.UnitTests/ClassEmitterTest.cs [hunk 1]
```diff
@@ -275,8 +275,8 @@ public void CreateEvent ()
     {
       var classEmitter = new CustomClassEmitter (Scope, "CreateEvent", typeof (object));
       CustomEventEmitter eventEmitter = classEmitter.CreateEvent ("Eve", EventKind.Instance, typeof (Func<string>));
-      eventEmitter.AddMethod.AddStatement (new ReturnStatement ());
-      eventEmitter.RemoveMethod.AddStatement (new ReturnStatement ());
+      eventEmitter.AddMethod?.AddStatement (new ReturnStatement ());
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection.CodeGeneration.UnitTests/ClassEmitterTest.cs:278]: NFO here.

File: Remotion/Core/Reflection.CodeGeneration/CustomEventEmitter.cs [hunk 2]
```diff
@@ -53,7 +53,7 @@ public CustomEventEmitter (CustomClassEmitter declaringType, string name, EventK
       declaringType.RegisterEventEmitter (this);
     }
 
-    public IMethodEmitter AddMethod
+    public IMethodEmitter? AddMethod
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection.CodeGeneration/CustomEventEmitter.cs:56]: not null. please also re-check usages and remove nfo / ?

File: Remotion/Core/Reflection.CodeGeneration/CustomEventEmitter.cs [hunk 3]
```diff
@@ -75,7 +75,7 @@ public IMethodEmitter AddMethod
       }
     }
 
-    public IMethodEmitter RemoveMethod
+    public IMethodEmitter? RemoveMethod
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection.CodeGeneration/CustomEventEmitter.cs:78]: not null, see above.

File: Remotion/Core/Reflection.CodeGeneration/CustomPropertyEmitter.cs [hunk 4]
```diff
@@ -70,7 +70,7 @@ public Type[] IndexParameters
       get { return _indexParameters; }
     }
 
-    public IMethodEmitter GetMethod
+    public IMethodEmitter? GetMethod
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection.CodeGeneration/CustomPropertyEmitter.cs:73]: interesting. that's just...bizarre. not sure though, if it's also a null-bug in the design.

File: Remotion/Core/Reflection/NullMethodInformation.cs [hunk 5]
```diff
@@ -28,22 +28,32 @@ namespace Remotion.Reflection
   /// </summary>
   public sealed class NullMethodInformation : IMethodInformation
   {
-    public string Name
+    public string? Name
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/NullMethodInformation.cs:31]: The null-implementations need a revamp. The fact that the not-nullablity has been ignored is a bug.

File: Remotion/Core/Reflection/NullMethodInformation.cs [hunk 6]
```diff
@@ -28,22 +28,32 @@ namespace Remotion.Reflection
...
+        return null;
+      }
+#nullable enable
     }
 
-    public ITypeInformation DeclaringType
+    public ITypeInformation? DeclaringType
     {
-      get { return null; }
+#nullable disable
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/NullMethodInformation.cs:43]: why?

File: Remotion/Core/Reflection/NullMethodInformation.cs [hunk 7]
```diff
@@ -28,22 +28,32 @@ namespace Remotion.Reflection
...
+      get 
+      {
+        return null;
+      }
+#nullable enable
     }
 
-    public ITypeInformation DeclaringType
+    public ITypeInformation? DeclaringType
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/NullMethodInformation.cs:41]: DeclaringType is marked as nullable on the interface

File: Remotion/Core/Reflection/NullPropertyInformation.cs [hunk 8]
```diff
@@ -18,6 +18,7 @@
 using System.Collections.Generic;
 using System.Reflection;
 
+#nullable disable
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/NullPropertyInformation.cs:21]: not disable

File: Remotion/Core/Reflection/NullPropertyInformation.cs [hunk 9]
```diff
@@ -27,12 +28,18 @@ public sealed class NullPropertyInformation : IPropertyInformation
...
+      get 
+      {
+        return null;
+      }
     }
 
     public ITypeInformation DeclaringType
     {
-      get { return null; }
+      get
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/NullPropertyInformation.cs:39]: no reformat

File: Remotion/Core/Reflection/PropertyInfoAdapter.cs [hunk 10]
```diff
@@ -163,16 +165,16 @@ public bool CanBeSetFromOutside
       return AttributeUtility.IsDefined<T> (_propertyInfo, inherited);
     }
 
-    public object GetValue (object instance, object[] indexParameters)
+    public object GetValue (object? instance, object[]? indexParameters)
     {
-      ArgumentUtility.CheckNotNull ("instance", instance);
+      ArgumentUtility.CheckNotNull ("instance", instance!);
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/PropertyInfoAdapter.cs:170]: null bug, see MethodInfoAdapter

File: Remotion/Core/Reflection/PropertyInfoAdapter.cs [hunk 11]
```diff
@@ -163,16 +165,16 @@ public bool CanBeSetFromOutside
...
 
       return _propertyInfo.GetValue (instance, indexParameters);
     }
 
-    public void SetValue (object instance, object value, object[] indexParameters)
+    public void SetValue (object? instance, object? value, object[]? indexParameters)
     {
-      ArgumentUtility.CheckNotNull ("instance", instance);
+      ArgumentUtility.CheckNotNull ("instance", instance!);
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/PropertyInfoAdapter.cs:177]: null bug, see MethodInfoAdapter

File: Remotion/Core/Reflection/PropertyInfoAdapter.cs [hunk 12]
```diff
@@ -232,7 +234,7 @@ public IEnumerable<IPropertyInformation> FindInterfaceDeclarations ()
       return _interfaceDeclarations.Value.ToArray();
     }
 
-    private IReadOnlyCollection<IPropertyInformation> FindInterfaceDeclarationsImplementation ()
+    private IReadOnlyCollection<IPropertyInformation?> FindInterfaceDeclarationsImplementation ()
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/PropertyInfoAdapter.cs:237]: not nullable

File: Remotion/Core/Reflection/PropertyInfoAdapter.cs [hunk 13]
```diff
@@ -109,7 +111,7 @@ private PropertyInfoAdapter (PropertyInfo propertyInfo)
           LazyThreadSafetyMode.ExecutionAndPublication);
 
       _interfaceDeclarations = new Lazy<IReadOnlyCollection<IPropertyInformation>> (
-          FindInterfaceDeclarationsImplementation,
+          FindInterfaceDeclarationsImplementation!,
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/PropertyInfoAdapter.cs:114]: why?

File: Remotion/Core/Reflection/MethodInfoAdapter.cs [hunk 14]
```diff
@@ -117,14 +119,14 @@ public ITypeInformation GetOriginalDeclaringType ()
       return AttributeUtility.GetCustomAttributes<T> (_methodInfo, inherited);
     }
 
-    public object Invoke (object instance, object[] parameters)
+    public object Invoke (object? instance, object?[]? parameters)
     {
-      ArgumentUtility.CheckNotNull ("instance", instance);
+      ArgumentUtility.CheckNotNull ("instance", instance!);
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/MethodInfoAdapter.cs:124]: null bug
question is, do we want to add a conditional argument check or let the MethodInfo handle this? need test either way.

File: Remotion/Core/Reflection/ReflectionBasedMemberInformationNameResolver.cs [hunk 15]
```diff
@@ -80,10 +81,12 @@ public string GetEnumName (Enum enumValue)
 
     private string GetPropertyNameInternal (IPropertyInformation propertyInformation)
     {
+#nullable disable
       return GetTypeName (propertyInformation.GetOriginalDeclaringType()) + "." + propertyInformation.Name;
+#nullable enable
     }
 
-    private string GetTypeNameInternal (ITypeInformation typeInformation)
+    private string? GetTypeNameInternal (ITypeInformation typeInformation)
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/ReflectionBasedMemberInformationNameResolver.cs:89]: That's a fun null bug :/

File: Remotion/Core/Reflection/NullMethodInformation.cs (hunk 5 above)
Comment (dkudernatsch) [reply Remotion/Core/Reflection/NullMethodInformation.cs:31]: What should i do with those, the interfaces declare something as not nullable and the implementation returns null. NFOing everything seems dangerous.

File: Remotion/Core/Reflection/PropertyInfoAdapter.cs (hunk 12 above)
Comment (dkudernatsch) [reply Remotion/Core/Reflection/PropertyInfoAdapter.cs:237]: FindDeclaringProperty may return null per interface definition

File: Remotion/Core/Reflection/NullMethodInformation.cs (hunk 5 above)
Comment (MichaelKetting) [reply Remotion/Core/Reflection/NullMethodInformation.cs:31]: Right now, that's what we will do. The fact that we need to change this to something better is the bugfix-todo&jira-ticket

File: Remotion/Core/Reflection/PropertyInfoAdapter.cs (hunk 12 above)
Comment (MichaelKetting) [reply Remotion/Core/Reflection/PropertyInfoAdapter.cs:237]: Hmm...I.C. 

Still wrong. If this were to occur, it would be a bug. It's not clear if the reflection implementation cannot trigger this scenario in the first place of if such members should be filtered. Fo now, we fix thsi with an NFO in the implementation and add a bugfix todo.

File: Remotion/Core/Reflection/MethodInfoAdapter.cs [hunk 16]
```diff
@@ -117,14 +119,15 @@ public ITypeInformation GetOriginalDeclaringType ()
       return AttributeUtility.GetCustomAttributes<T> (_methodInfo, inherited);
     }
 
-    public object Invoke (object instance, object[] parameters)
+    public object Invoke (object? instance, object?[]? parameters)
     {
-      ArgumentUtility.CheckNotNull ("instance", instance);
+      //TODO RM-7432: Implementation requires not null on nullable parameter
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection/MethodInfoAdapter.cs:124]: something got lost somewhere in the comments, don't know where, don't know how, don't case.

here and in property info (GetValue, SetValue): the instance can be null, the null-check is wrong. -> otherwise, we don't support statics and that's just wrong.

File: Remotion/Core/Reflection/ITypeInformation.cs [hunk 17]
```diff
@@ -249,7 +249,7 @@ public interface ITypeInformation : IMemberInformation
     /// or <see langword="null"/> if the current type is not an array or a pointer, or is not passed by reference, 
     /// or represents a generic type or a type parameter in the definition of a generic type or generic method.
     /// </returns>
-    [CanBeNull]ITypeInformation GetElementType ();
+    [CanBeNull] ITypeInformation? GetElementType ();
```
Comment (iam-mholle) [inline Remotion/Core/Reflection/ITypeInformation.cs:252]:
```suggestion
    [CanBeNull]ITypeInformation? GetElementType ();
```

File: Remotion/Core/Reflection/ReflectionBasedMemberInformationNameResolver.cs [hunk 18]
```diff
@@ -67,7 +67,6 @@ public string GetPropertyName (IPropertyInformation propertyInformation)
     public string GetTypeName (ITypeInformation typeInformation)
     {
       ArgumentUtility.CheckNotNull ("typeInformation", typeInformation);
-
```
Comment (iam-mholle) [inline Remotion/Core/Reflection/ReflectionBasedMemberInformationNameResolver.cs:70]: Unnecessary change.

File: Remotion/Core/Reflection/ReflectionExtensions.cs [hunk 19]
```diff
@@ -51,7 +51,7 @@ public static bool IsOriginalDeclaration (this IMemberInformation memberInfo)
     }
 
     [NotNull]
-    public static Type ConvertToRuntimeType (this ITypeInformation typeInformation)
+    public static Type? ConvertToRuntimeType (this ITypeInformation typeInformation)
```
Comment (iam-mholle) [inline Remotion/Core/Reflection/ReflectionExtensions.cs:54]: That's a bit contradictory 😉

File: Remotion/Core/Reflection.CodeGeneration/DPExtensions/ExpressionReference.cs [hunk 20]
```diff
@@ -65,7 +65,7 @@ private LocalReference CreateLocal (ILGenerator gen)
       return local;
     }
 
-    public override void StoreReference (ILGenerator gen)
+    public override void StoreReference (ILGenerator? gen)
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply Remotion/Core/Reflection.CodeGeneration/DPExtensions/ExpressionReference.cs:68]: so, notnull and a nfo in tests?

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Core/Reflection.CodeGeneration/DPExtensions/ExpressionReference.cs:68]: yes. impl is not supported, value in test is probably a stub (I'm assuming) with ignored argument values. in generel: wenn the test passes dmmy values because IgnoreArguments, those are not indicative of the signatures nullablity.

File: Remotion/Core/Reflection.CodeGeneration/CustomPropertyEmitter.cs (hunk 4 above)
Comment (MichaelKetting) [reply Remotion/Core/Reflection.CodeGeneration/CustomPropertyEmitter.cs:73]: not with the current implementation differences, we can't support a symmetrical API that's also safe. this difference might well be a bug and the get and set method be not-null. it's also possible that there's a subtle difference between event accessors and property accessors. either way, we can't repair this here and we can't fudge on the api without making it buggier. so, yes, prperty and event will be different until someone figures out if this is correct and fixes it in the correct order. given that Reflection.CodeGen is a legacy assembly, that won't happen anytime soon :D
